pubtools/_pulp/hooks.py

```diff
@@ -14,4 +14,13 @@ def task_pulp_flush():
     Pulp content may have recently changed."""
 
 
+@hookspec
+def pulp_item_finished(item_metadata=None):  # pylint: disable=unused-argument
```

Comment:
The PostPushActions code is missing right now, but based on context I see what this is going for.

I've got some issues with the data currently being passed to this hook: the PulpPushItem classes are meant to be internal, they are not designed as public API.  But as soon as we start passing them into this hook then we are "locked in", we can't refactor them any more without breaking whoever's on the other end of the hook.  We need to be strict about what we're producing for this hook in order to avoid issues.  Some of the fields you're exporting now (like upload_key) were absolutely intended to be just "this is what works for the code internally" and it is very risky to let other libraries start depending on them.

Is it possible to satisfy your use-case by yielding only public classes from pushsource and/or pubtools-pulplib libraries?  I'm guessing that it might be - you're interested in cdn_path but that's a documented field on FileUnit at https://release-engineering.github.io/pubtools-pulplib/api/files.html#pubtools.pulplib.FileUnit.cdn_path .

What do you think of this alternative signature then:

```python
def pulp_item_finished(pushitem, pulp_units):
```

...where:

- pushitem is an instance of [pushsource.PushItem](https://release-engineering.github.io/pushsource/model/base.html#pushsource.PushItem)
- pulp_units is a list of [pulplib.Unit](https://release-engineering.github.io/pubtools-pulplib/api/pulpcore.html#pubtools.pulplib.Unit) (list keeps the possibility open of one PushItem generating zero units or more than one unit).  These are the units created or updated in Pulp for that item.
- Obviously it would be the concrete subclasses like FilePushItem, FileUnit depending on the specific content type.

Reply:
yes, I was particularly interested in cdn_path